### PR TITLE
[SPARK-37561][SQL] Avoid loading all functions when obtaining hive's DelegationToken

### PR DESCRIPTION
**Pull request by cxzl25**
### What changes were proposed in this pull request?
Use `Hive.getWithoutRegisterFns` instead of `Hive.get` to avoid loading all permanent functions from the hive meta store.

### Why are the changes needed?
At present, when obtaining the delegationToken of hive, all functions will be loaded.
This is unnecessary, it takes time to load the function, and it also increases the burden on the hive meta store.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
manual test

### CODE REVIEW

**Comment by cxzl25** [comment]
getDelegationToken_load_functions

**Comment by HyukjinKwon** [comment]
cc @sunchao

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/security/HiveDelegationTokenProvider.scala [hunk 1]
```diff
@@ -99,7 +99,12 @@ private[spark] class HiveDelegationTokenProvider
         s"$principal at $metastoreUri")
 
       doAsRealUser {
-        val hive = Hive.get(conf, classOf[HiveConf])
+        val hive = conf match {
+          case hiveConf: HiveConf =>
+            Hive.getWithoutRegisterFns(hiveConf)
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/security/HiveDelegationTokenProvider.scala:104]: we will have to nunuc https://github.com/apache/spark/blob/116255d10989fd87d1eb3347bb279df207b1b5c4/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L204-L211 approach to allow other versions of hive when buidling. cc @AngersZhuuuu FYI

Comment (cxzl25) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/security/HiveDelegationTokenProvider.scala]: Ok, I thought at first that code that has nothing to do with `HiveClientImpl` does not need to be compatible with lower versions. I also use reflection to solve compatibility problems.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/security/HiveDelegationTokenProvider.scala [hunk 2]
```diff
@@ -99,7 +99,18 @@ private[spark] class HiveDelegationTokenProvider
         s"$principal at $metastoreUri")
 
       doAsRealUser {
-        val hive = Hive.get(conf, classOf[HiveConf])
+        val hive = conf match {
+          case hiveConf: HiveConf =>
+            try {
```
Comment (sunchao) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/security/HiveDelegationTokenProvider.scala:104]: can we reduce the code duplication by moving `HiveClientImpl.getHive` to its companion object and then call the method here?

Comment (cxzl25) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/security/HiveDelegationTokenProvider.scala]: done

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 3]
```diff
@@ -1287,4 +1276,26 @@ private[hive] object HiveClientImpl extends Logging {
     }
     hiveConf
   }
+
+  /**
+   * Initialize Hive through Configuration.
+   * first try to use getWithoutRegisterFns to initialize to avoid loading all functions,
```
Comment (sunchao) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:1282]: nit: `first` -> `First`

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 4]
```diff
@@ -1287,4 +1276,26 @@ private[hive] object HiveClientImpl extends Logging {
...
+          classOf[Hive].getMethod("getWithoutRegisterFns", classOf[HiveConf])
+            .invoke(null, conf).asInstanceOf[Hive]
+        } catch {
+          // SPARK-37069: not all Hive versions have the above method (e.g., Hive 2.3.9 has it but
+          // 2.3.8 don't), therefore here we fallback when encountering the exception.
+          case _: NoSuchMethodException =>
+            Hive.get(hiveConf)
+        }
+      case _ =>
+        Hive.get(conf, classOf[HiveConf])
```
Comment (sunchao) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:1298]: Maybe we can also call `getWithoutRegisterFns` for this path? by first converting the `Configuration` to a `HiveConf`.

**Review by sunchao — APPROVED** [review]
LGTM (pending CI).

**Comment by sunchao** [comment]
Merged, thanks!

**Pull request closed without merging**